### PR TITLE
feat: nth fibonacci number using memoization.

### PR DESCRIPTION
**Pull request by pradyyadav**
#### Description of Change
Added the implementation of Memoization to get the nth Fibonacci number.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass.
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

I noticed no file explaining the memoization technique to get the nth Fibonacci number, which is the key aspect of dynamic programming.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1534"></a>

### CODE REVIEW

**Comment by foo290** [comment]
Please refer to the [contribution guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md) for coding format and docs and consider adding detailed documentation for the code. 😃

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Yep, as @foo290 said, code is not up to the [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md).
Please read them **carefully** and **follow them**.

**Comment by pradyyadav** [comment]
Coding format and docs are modified according to repository standards. @Panquesito7 @foo290

File: dynamic_programming/fibonacci_memoization.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
+/**
+ * @file
+ * @brief Implementation of [nth Fibonacci problem using 
+ * memoization](https://medium.com/@andrewsouthard1/dynamic-programming-recursion-memoization-and-bottom-up-algorithms-61c882d1c7e)
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:4]:
```suggestion
 * @brief Implementation of [nth Fibonacci problem](https://medium.com/@andrewsouthard1/dynamic-programming-recursion-memoization-and-bottom-up-algorithms-61c882d1c7e) using memoization
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * @details
+ * nth Fibonacci number using memoization is used as a dynamic programming concept that
+ * works in \f$O(n)\f$ time and works in \f$O(n)\f$ space.
+ * Other solution is bottom up approach which uses \f$O(1)\f$ space. 
+ * @author [Pradyumna Yadav](https://github.com/pradyyadav)
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for io operations
+#include <unordered_map>    /// for std::unordered_map
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:14]:
```suggestion
#include <cassert>            /// for assert
#include <iostream>          /// for io operations
#include <unordered_map>    /// for std::unordered_map
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace fibonacci_memoization
+ * @brief Functions for [nth Fibonacci
+ * using memoization](https://medium.com/@andrewsouthard1/dynamic-programming-recursion-memoization-and-bottom-up-algorithms-61c882d1c7e)
+ * algorithm
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:25]:
```suggestion
 * @brief Functions for the [nth Fibonacci](https://medium.com/@andrewsouthard1/dynamic-programming-recursion-memoization-and-bottom-up-algorithms-61c882d1c7e) using memoization implementation
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+ */
+namespace fibonacci_memoization {
+/**
+ * @brief The function that implements the nth Fibonacci number problem using
+ * memoization
+ * @param n The index of the number in a fibonacci series.
+ * @param map_elements A map that contains previously encountered index-value pair.
+ * @returns Element at the nth position in fibonacci series.
+ */
+std::uint32_t nthFibonacci(int n, auto &map_elements)
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:35]:
```suggestion
uint32_t nthFibonacci(int n, auto &map_elements)
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+ * using memoization](https://medium.com/@andrewsouthard1/dynamic-programming-recursion-memoization-and-bottom-up-algorithms-61c882d1c7e)
+ * algorithm
+ */
+namespace fibonacci_memoization {
+/**
+ * @brief The function that implements the nth Fibonacci number problem using
+ * memoization
+ * @param n The index of the number in a fibonacci series.
+ * @param map_elements A map that contains previously encountered index-value pair.
+ * @returns Element at the nth position in fibonacci series.
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:33]:
```suggestion
 * @returns Element at the nth position in the Fibonacci series.
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+ * @namespace fibonacci_memoization
+ * @brief Functions for [nth Fibonacci
+ * using memoization](https://medium.com/@andrewsouthard1/dynamic-programming-recursion-memoization-and-bottom-up-algorithms-61c882d1c7e)
+ * algorithm
+ */
+namespace fibonacci_memoization {
+/**
+ * @brief The function that implements the nth Fibonacci number problem using
+ * memoization
+ * @param n The index of the number in a fibonacci series.
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:31]:
```suggestion
 * @param n The index of the number in a Fibonacci series.
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+}
+}  // namespace fibonacci_memoization
+}  // namespace dynamic_programming
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:53]:
```suggestion
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    // Test 1
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:54]:
```suggestion
    // 1st test
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    // 7 return 13.
+    int num1 = 7;
+    std::unordered_map<int,uint32_t> map_elements1;
+    std::cout << "Test 1... ";
+    assert(dynamic_programming::fibonacci_memoization::nthFibonacci(num1, map_elements1) == 13); // 7th fibonacci number is 13.
+    std::cout<< "passed" << std::endl;
+
+
+    // Test 2
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:63]:
```suggestion
    // 2nd test
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    // 45 return 1134903170.
+    int num2 = 45;
+    std::unordered_map<int,uint32_t> map_elements2;
+    std::cout << "Test 2... ";
+    assert(dynamic_programming::fibonacci_memoization::nthFibonacci(num2, map_elements2) == 1134903170); // 45th fibonacci number is 1134903170.
+    std::cout<< "passed" << std::endl;
+
+    // Test 3
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:71]:
```suggestion
    // 3rd test
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    // 15 return 610.
+    int num3 = 15;
+    std::unordered_map<int,uint32_t> map_elements3;
+    std::cout << "Test 3... ";
+    assert(dynamic_programming::fibonacci_memoization::nthFibonacci(num3, map_elements3) == 610); // 15th fibonacci number is 610.
+    std::cout<< "passed" << std::endl;
+
+    // Test 4
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:79]:
```suggestion
    // 4th test
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();  // run self-test implementations
+    return 0;
+}
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:95]:
```suggestion
}

```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 👍

File: dynamic_programming/fibonacci_memoization.cpp [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+ * @brief The function that implements the nth Fibonacci number problem using
+ * memoization
+ * @param n The index of the number in a fibonacci series.
+ * @param map_elements A map that contains previously encountered index-value pair.
+ * @returns Element at the nth position in fibonacci series.
+ */
+std::uint32_t nthFibonacci(int n, auto &map_elements)
+{
+    if (n <= 1)  // The numbers at index 0 and 1 are also 0 and 1.
+        return n;
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:38]:
```suggestion
    if (n <= 1) {  // The numbers at index 0 and 1 are also 0 and 1.
        return n;
    }
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1534/checks?check_run_id=3089146663) warnings.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1534/checks?check_run_id=3156953517) warnings.

**Comment by pradyyadav** [comment]
@Panquesito7 I am unable to figure out what is wrong with my code, the checks keeps failing in code formatter section.

File: dynamic_programming/fibonacci_memoization.cpp [hunk 14]
```diff
@@ -0,0 +1,96 @@
+/**
+ * @file
+ * @brief Implementation of [nth Fibonacci problem](https://medium.com/@andrewsouthard1/dynamic-programming-recursion-memoization-and-bottom-up-algorithms-61c882d1c7e) using memoization
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:3]:
```suggestion
 * @brief Implementation of [nth Fibonacci problem](https://medium.com/@andrewsouthard1/dynamic-programming-recursion-memoization-and-bottom-up-algorithms-61c882d1c7e)
 * using memoization
```

File: dynamic_programming/fibonacci_memoization.cpp [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+ * @brief Functions for the [nth Fibonacci](https://medium.com/@andrewsouthard1/dynamic-programming-recursion-memoization-and-bottom-up-algorithms-61c882d1c7e) using memoization implementation
+ */
+namespace fibonacci_memoization {
+/**
+ * @brief The function that implements the nth Fibonacci number problem using memoization.
+ * @param n The index of the number in a Fibonacci series.
+ * @param term
+ * @returns Element at the nth position in the Fibonacci series.
+ */
+long long int nthFibonacci(int n, int term[])
```
Comment (Panquesito7) [inline dynamic_programming/fibonacci_memoization.cpp:31]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). **Let us know if you need any help**. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: dynamic_programming/fibonacci_memoization.cpp (hunk 15 above)
Comment (pradyyadav) [reply dynamic_programming/fibonacci_memoization.cpp:31]: Thank you, actually when I am using command: ```clang-tidy --fix --quiet -p build dynamic_programming/fibonacci_memoization.cpp --``` inside my local repo folder, it is throwing this error: ```clang-tidy: The term 'clang-tidy' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.``` I know it is due to some path error. Can you please tell me what's wrong here.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1534/checks?check_run_id=3654747340) warnings.

**Comment by pradyyadav** [comment]
> Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1534/checks?check_run_id=3654747340) warnings.
> 

`error: assignment of read-only location` which is obviously due to `const std::vector<uint32_t> &term`. 

When removing `const`, the clang-tidy throws `error: non-const reference parameter 'term', make it const or use a pointer [google-runtime-references,-warnings-as-errors]
uint32_t nthFibonacci(const uint32_t &n, std::vector<uint32_t> &term).` What should I do ?

**Comment by Panquesito7** [comment]
> > Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1534/checks?check_run_id=3654747340) warnings.
> 
> `error: assignment of read-only location` which is obviously due to `const std::vector<uint32_t> &term`.
> 
> When removing `const`, the clang-tidy throws `error: non-const reference parameter 'term', make it const or use a pointer [google-runtime-references,-warnings-as-errors] uint32_t nthFibonacci(const uint32_t &n, std::vector<uint32_t> &term).` What should I do ?

Try to remove the `&` for the `term` variable and see if everything works properly. 🙂

**Pull request closed without merging**